### PR TITLE
routing: Add Validation for Decoding Blinded Paths

### LINKED ISSUE #8137: [bug]: blinded forwarding: hop encoding / payload decoding inconsistency
To support blinded forwarding in the future, https://github.com/lightningnetwork/lnd/commit/940b491051d2f61225424551e52d8f2118e3d2cf started omitting the amount and timelock TLVs in `Hop.PackHopPayload` when those values are 0.  However, no matching change was made to the payload decoding logic in `NewPayloadFromReader` and `ValidateParsedPayloadTypes`.  As a result, onion payloads along a blinded route will fail to decode due to missing amount or timelock TLVs.

Although blinded forwarding is still a [WIP](https://github.com/lightningnetwork/lnd/issues/7298), it would be good to fix this inconsistency at least for the sake of our `FuzzPayload` test, which currently fails with the following input:

```shell
$ cd htlcswitch/hop
$ cat testdata/fuzz/FuzzPayload/2892b85ab66a5aec6ae8f500e357b887f37b1041fd71186d4e541127e01e49b0
go test fuzz v1
[]byte("\x02\x00\x04\x0200")

$ go test -run=FuzzPayload/2892b85ab66a5aec6ae8f500e357b887f37b1041fd71186d4e541127e01e49b0
--- FAIL: FuzzPayload (0.00s)
    --- FAIL: FuzzPayload/2892b85ab66a5aec6ae8f500e357b887f37b1041fd71186d4e541127e01e49b0 (0.00s)
        fuzz_test.go:122: 
                Error Trace:    htlcswitch/hop/fuzz_test.go:122

                Error:          Received unexpected error:
                                onion payload for final hop omitted record with type 2
                Test:           FuzzPayload/2892b85ab66a5aec6ae8f500e357b887f37b1041fd71186d4e541127e01e49b0
FAIL
exit status 1
FAIL    github.com/lightningnetwork/lnd/htlcswitch/hop  0.007s
```

CC: @carlaKC

### PR DESCRIPTION
**Pull request by carlaKC**
Fixes #8137 and expands validation of blinded routes. 

As outlined in the original issue, the fuzzer ran into the following problems: 
- We allow creation of blinded route payloads in `PackHopPayload`.
- We do not allow deserialization of the payloads in `NewPayloadFromReader` because LND doesn't support forwarding blinded routes. 

This PR updates validation to be consistent across all validation functions. 

A few of the commits in the beginning of the PR are just fixing up test data to be more accurate before stricter validation is implemented. They're included separately to make sure that they pass with the new and old code.

### CODE REVIEW

**Comment by carlaKC** [comment]
@morehouse opening this up early for a first look, specifically want to make sure that I'm grasping the fuzz test properly: 
* With the introduction of blinded routes, we use external context (ie the sphinx packet) to determine whether a hop belongs to the final node (previously we used the `nextChanID` context that is provided within the payload).
* We need to tell the fuzzer this external context so that is can create and test valid payloads for that specific context (previously we were just inferring from whatever value the fuzzer gave us for `nextChanID`)

To do this, I've split fuzzing up into final and non-final tests, but I'm not sure that it's the best-practice way of doing things? I could also pull a byte off the fuzzing data to decide this context?

File: routing/route/route.go [hunk 1]
```diff
@@ -193,8 +200,28 @@ func (h *Hop) PackHopPayload(w io.Writer, nextChanID uint64,
...
+		return fmt.Errorf("%w: amount to forward: %v", err,
+			h.AmtToForward)
+	}
+
+	if err := optionalBlindedField(
+		h.OutgoingTimeLock == 0, isBlinded, lastHop,
+	); err != nil {
+		return fmt.Errorf("%w: outgoing timelock: %v", err,
+			h.OutgoingTimeLock)
+	}
```
Comment (morehouse) [inline routing/route/route.go:221]: For the fuzz test to stop failing, we need a similar check on the decode side (`NewPayloadFromReader`).

File: htlcswitch/hop/payload.go [hunk 2]
```diff
@@ -256,33 +256,48 @@ func ValidateParsedPayloadTypes(parsedTypes tlv.TypeMap,
...
+	needFwdInfo := isFinalHop || !hasEncryptedData
+
+	// No blinded hops should have a next hop specified, and only the final
+	// hop in a cleartext route should exclude it.
+	needNextHop := !(hasEncryptedData || isFinalHop)
 
+	switch {
 	// All hops must include an amount to forward.
-	case !hasAmt:
+	case needFwdInfo && !hasAmt:
```
Comment (morehouse) [inline htlcswitch/hop/payload.go:271]: There's a corner case here.  `hasAmt` will be true if the TLV is present but 0.

The final hop fuzz test still fails for this input:  `[]byte("\x02\x00\x04\x00")`.  Decoding succeeds because the TLV is present but 0.  Then encoding fails because the amount is 0.

I'm not sure if we should fail to decode 0 amount or locktime here or if we should check that at a higher level somewhere (maybe we already do that?).  Failing during decoding would make the fuzz test start passing.  Otherwise, if we fail at a higher level, we'll need to except this corner case in the fuzz test.

Comment (carlaKC) [reply htlcswitch/hop/payload.go:271]: > I'm not sure if we should fail to decode 0 amount or locktime here or if we should check that at a higher level somewhere (maybe we already do that?).

Yeah, we check that the HTLC's amount/expiry are sane later on. 

> Otherwise, if we fail at a higher level, we'll need to except this corner case in the fuzz test.

Somewhat torn here. I don't think that failing during decoding for a TLV that is present but zero is the right approach (value semantics don't belong in decoding),  but it seems like a shame to exclude all of these cases from fuzzing - iiuc we'd need to not fail on `ErrMissingField` because `amt`/`expiry` or even `chanid` could be present but zero?

File: htlcswitch/hop/payload_test.go [hunk 3]
```diff
@@ -356,6 +356,51 @@ var decodePayloadTests = []decodePayloadTest{
...
+		payload: []byte{
+			// amount
+			0x02, 0x00,
+			// cltv
+			0x04, 0x00,
+			// encrypted data
+			0x0a, 0x03, 0x03, 0x02, 0x01,
+		},
+		shouldHaveEncData: true,
+	},
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply htlcswitch/hop/payload_test.go:371]: Shouldn't be included for the test cases below - since they're error tests we never get to the point where we check the value of the payload so it had no impact on the test, will delete.

File: htlcswitch/hop/payload.go [hunk 4]
```diff
@@ -248,42 +249,55 @@ func NewCustomRecords(parsedTypes tlv.TypeMap) record.CustomSet {
...
 		}
 
 	// All hops must include a cltv expiry.
-	case !hasLockTime:
+	case needFwdInfo && !hasLockTime:
 		return ErrInvalidPayload{
 			Type:      record.LockTimeOnionType,
 			Violation: OmittedViolation,
 			FinalHop:  isFinalHop,
 		}
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply htlcswitch/hop/payload.go:285]: Added in these two cases + manually tested `FuzzPayload` with `[]byte("\x04\x0200\n\x00")` to make sure this is addressed 👍

**Comment by morehouse** [comment]
Code looks good and the fuzz test seems happy again (mostly).  There's another small change that needs to be made to the fuzz tests to accommodate the route blinding stuff, but it doesn't belong in this PR.  I'll send a separate PR once this one is merged.

I'll run the fuzz tests overnight to check for any other problems, and do a final review tomorrow if everything passes.

There' a couple small unresolved comments, but nothing blocking.

**Review by morehouse — APPROVED** [review]

**Comment by saubyk** [comment]
cc: @ziggie1984 for review

File: routing/route/route_test.go [hunk 5]
```diff
@@ -160,18 +160,113 @@ func TestAMPHop(t *testing.T) {
...
+	}{
+		{
+			name: "introduction point with next channel",
+			hop: Hop{
+				EncryptedData: []byte{1, 2, 3},
+				BlindingPoint: testPubKey,
+			},
+			nextChannel: 1,
+			isFinal:     false,
+			err:         ErrIncorrectField,
```
Comment (ziggie1984) [inline routing/route/route_test.go:183]: Nit: `ErrIncorrectField` => `ErrOmittedField` because technically the tlv field is correct but just not allowed?

File: htlcswitch/hop/fuzz_test.go [hunk 6]
```diff
@@ -122,15 +122,26 @@ func fuzzPayload(f *testing.F, finalPayload bool) {
...
+		// record, since validation is done at a later stage. Do
+		// not report a bug for this case.
+		case errors.Is(err, route.ErrAMPMissingMPP):
 			return
+
+		// PackHopPayload will not encode regular payloads or final
+		// hops in blinded routes that do not have an amount or expiry
+		// TLV set. However, NewPayloadFromReader will allow creation
+		// of payloads where these TLVs are present, but they have
+		// zero values because validation is done at a later stage.
```
Comment (ziggie1984) [inline htlcswitch/hop/fuzz_test.go:138]: Q: Do we validate the zero values also when forwarding an HTLC or what do you mean exactly by:

```validation is done at a later stage ?```

**Review by ziggie1984 — COMMENTED** [review]
Great addition of validation logic when creating/parsing the payload data.

PR looks really good 👏, just had some minor nits 

For me personally I found the commit splitting a bit too fine granular because all these small testcase commits in the beginning were clear to me after I reviewed the commit where the validation logic was added.

File: routing/route/route_test.go [hunk 7]
```diff
@@ -197,6 +197,7 @@ func TestPayloadSize(t *testing.T) {
 			PubKeyBytes:      testPubKeyBytes,
 			AmtToForward:     1200,
 			OutgoingTimeLock: 700000,
+			ChannelID:        51784534844,
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply routing/route/route_test.go:295]: > because we are building the route in reverse order I think this can never happen in real life

yeah, bad test values - will update in next push (though it doesn't _really_ matter in the context of this test).

> I would have liked this commit to be part of the commit where the validation is added, but that's just my personal taste

Reasoning for this is to demonstrate that these test values are valid before the stricter validation and after. I like having the updates to the tests in separate commits so that we're changing one thing at a time (rather than updating it with the validation, and then perhaps hitting a different code path by mistake).

> otherwise pretty hard to reason about this addition

Is there something that I could add to the commit message to clarify this? Did try to explain the reasoning in there.

File: htlcswitch/hop/iterator_test.go [hunk 8]
```diff
@@ -65,14 +65,16 @@ func TestSphinxHopIteratorForwardingInstructions(t *testing.T) {
 			},
 			expectedFwdInfo: expectedFwdInfo,
 		},
-		// A TLV payload, we can leave off the action as we'll always
-		// read the cid encoded.
+		// A TLV payload, which includes the sphinx action as
+		// cid may be zero for blinded routes (thus we require the
+		// action to signal that we are at the final hop).
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply htlcswitch/hop/iterator_test.go:70]: > Q: While looking at this Action parameter, it seems to me that it was not used before this PR, strange or ?

Yeah also odd to me - I don't know why we previously did not use the `Action` param tbh.

File: htlcswitch/hop/fuzz_test.go (hunk 6 above)
Comment (carlaKC) [reply htlcswitch/hop/fuzz_test.go:138]: Just referring to the validation that we do when adding the HTLC to our channel (which disallows zero values). Copied the phrasing from the comment for AMP above.

File: routing/route/route_test.go (hunk 5 above)
Comment (carlaKC) [reply routing/route/route_test.go:183]: Omitted means that we left the field out when we should have included it, which is different to the case where we included a field that shouldn't be there? Updated this to `ErrUnexpectedField`, hope that's clearer?

File: htlcswitch/hop/fuzz_test.go (hunk 6 above)
Comment (ziggie1984) [reply htlcswitch/hop/fuzz_test.go:138]: Cool, I haven't looked into the Forwarding Blinded Payments PR but will we allow Onions which have the TLV value set but the for example the value is set to 0 pass through ?

Comment (carlaKC) [reply htlcswitch/hop/fuzz_test.go:138]: >  Onions which have the TLV value set but the for example the value is set to 0 pass through ?

By the time we get to the outgoing link, we've calculated the amount that we _actually_ need to forward from the encrypted data.

File: htlcswitch/hop/payload_test.go [hunk 9]
```diff
@@ -348,6 +348,81 @@ var decodePayloadTests = []decodePayloadTest{
...
+	{
```
Comment (ziggie1984) [inline htlcswitch/hop/payload_test.go:364]: Nit: While checking the code coverage, I found that the new validation where an intermediate hop has a zero chanID is not part of this test, but maybe it's tested in the routing pkg. 

Screenshot 2023-12-16 at 11 05 42

**Review by ziggie1984 — APPROVED** [review]
Very clean and great PR: LGTM ⚡️

File: htlcswitch/hop/iterator_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (carlaKC) [reply htlcswitch/hop/iterator_test.go:70]: Referring to the general mechanism of using the action rather than the specifics of this test case. 

Would "signal that we are" -> "signal whether we are" make it clearer?

**Comment by carlaKC** [comment]
[Pushed](https://github.com/lightningnetwork/lnd/compare/eec6795848d7745cf245b18ea39cb1a0457123ca..31d4242b6d636520357b5f6683ea5142017394b3) comment change + additional test case for validation.
